docs/advanced_concepts.md

```diff
@@ -11,6 +11,48 @@ report, and adding animations.
    <https://discuss.streamlit.io/>`_.
 ```
 
+## Batch elements and input widgets
+
+In Streamlit, every widget interaction causes a rerun of the app. However,
+there are times when you might want to interact with a couple of widgets and
+submit those interactions while triggering a single re-run of the app.
+
+Using `st.form` you can batch input widgets together and along with
+`st.form_submit_button` submit the state inside these widgets with the click
+of a single button.
+
+```python
+# Forms can be declared using the 'with' syntax
+with st.form(key='my_form'):
+    text_input = st.text_input(label='Enter your name')
+    submit_button = st.form_submit_button(label='Submit')
+```
+
+```python
+# Alternative syntax, declare a form and use the returned object
+form = st.form(key='my_form')
```

Comment:
Not something to hold up this PR, but I wonder if we should start to be more opinionated about the documentation, such that we always use the `with` syntax where available.